eclair-core/src/main/scala/fr/acinq/eclair/payment/Invoice.scala

```diff
@@ -41,7 +42,7 @@ trait Invoice {
 
 object Invoice {
   /** An extra edge that can be used to pay a given invoice and may not be part of the public graph. */
-  case class ExtraEdge(sourceNodeId: PublicKey,
+  case class ExtraEdge(sourceNodeId: Either[ShortChannelIdDir, PublicKey],
```

Comment:
I believe we could revert many changes in this PR (and thus greatly simplify the code) if we avoided that specific change. Those extra edges are only necessary for payments, and not for the offer, is that correct?

If so, I think that the flow should look like this:

- use the bolt 12 offer to obtain a bolt 12 invoice (in the `invoice_request` step)
- for every blinded path in the bolt 12 invoice, if some of them use an `sciddir`, resolve them to a `node_id`
  - if there are some blinded paths that we cannot resolve to a `node_id`, drop them from the invoice before passing it to the payment actors, since it would fail route calculation
  - if there are no usable blinded paths after that filtering step, we can fail even before going through the payment actors

This way the `ExtraEdge` really stays a `sourceNodeId -> targetNodeId`, thanks to one pre-processing step on the invoice. I think the code would be easier to follow that way, we would "hide" the optimization of using an `sciddir` outside of the payment handling and route calculation components.

What do you think? Was this something you tried?